assets/js/frontend/add-to-cart.js

```diff
@@ -76,33 +76,35 @@ jQuery( function( $ ) {
 			});
 
 			// Trigger event.
-			$( document.body ).trigger( 'adding_to_cart', [ $thisbutton, data ] );
-
-			e.data.addToCartHandler.addRequest({
-				type: 'POST',
-				url: wc_add_to_cart_params.wc_ajax_url.toString().replace( '%%endpoint%%', 'add_to_cart' ),
-				data: data,
-				success: function( response ) {
-					if ( ! response ) {
-						return;
-					}
-
-					if ( response.error && response.product_url ) {
-						window.location = response.product_url;
-						return;
-					}
-
-					// Redirect to cart option
-					if ( wc_add_to_cart_params.cart_redirect_after_add === 'yes' ) {
-						window.location = wc_add_to_cart_params.cart_url;
-						return;
-					}
-
-					// Trigger event so themes can refresh other areas.
-					$( document.body ).trigger( 'added_to_cart', [ response.fragments, response.cart_hash, $thisbutton ] );
-				},
-				dataType: 'json'
-			});
+			var valid = $( document.body ).triggerHandler( 'adding_to_cart', [ $thisbutton, data ] );
+
+			if( valid !== false ) {
+				e.data.addToCartHandler.addRequest({
+					type: 'POST',
+					url: wc_add_to_cart_params.wc_ajax_url.toString().replace( '%%endpoint%%', 'add_to_cart' ),
+					data: data,
+					success: function( response ) {
+						if ( ! response ) {
+							return;
+						}
+
+						if ( response.error && response.product_url ) {
+							window.location = response.product_url;
+							return;
+						}
+
+						// Redirect to cart option
+						if ( wc_add_to_cart_params.cart_redirect_after_add === 'yes' ) {
+							window.location = wc_add_to_cart_params.cart_url;
+							return;
+						}
+
+						// Trigger event so themes can refresh other areas.
+						$( document.body ).trigger( 'added_to_cart', [ response.fragments, response.cart_hash, $thisbutton ] );
```

Comment:
When this event is not fired the button remains in the AJAX state eternally. Perhaps we should add an error event trigger as well, which can call `updateState`. We should also add a failure handler that does the same if we do, which will fix a bug here too!

Reply:
This can be resolved by firing this _before_ the ajax state is set on the button. Or alternatively, as you say we can trigger a `failure` event and attach `update_button()` to it.

Reply:
I've added a `$( document.body ).trigger( 'ajax_request_not_sent.adding_to_cart', [ false, false, $thisbutton ] );` for when the triggerHandler is set to false. And attached `updateButton` to it.

Reply:
What about just `adding_to_cart_prevented`? The dot notation isn't really consistent with any other event in the cart.